sfepy/discrete/problem.py

```diff
@@ -985,18 +986,27 @@ def get_ebc_indices(self):
     def set_conf_solvers(self, conf_solvers=None, options=None):
         """
         Choose which solvers should be used. If solvers are not set in
-        `options`, use first suitable in `conf_solvers`.
+        `options`, use the ones named `ls`, `nls` or `ts`. If such solvers
+        do not exist, us the firsts suitable in `conf_solvers`.
         """
         conf_solvers = get_default(conf_solvers, self.conf.solvers)
         self.solver_confs = {}
+
         for key, val in six.iteritems(conf_solvers):
             self.solver_confs[val.name] = val
 
         def _find_suitable(prefix):
+            cands = []
             for key, val in six.iteritems(self.solver_confs):
                 if val.kind.find(prefix) == 0:
-                    return val
-            return None
+                    if val.name == prefix[:-1]:
+                        return val
+                    else:
+                        cands.append(val)
+            if len(cands) > 0:
+                return cands[0]
+            else:
+                return None
```

Comment:
What is the reason of this change? I mean, why creating a list, when the first item is always used.

Reply:
This change ensures to select either `ls`/`nls`/`ts`, as required by `_get_solver_conf()`, if it exists, or the first linear/nonlinear/time solver in the created list.

Examples:
 required 'ls', defined solvers ['ls3', 'ls2', 'ls']  -> 'ls'
 required 'ls', defined solvers ['ls3', 'ls2', 'ls1']  -> 'ls3'

Reply:
OK, so you probably do not need to append to the list - just replacing the first item would be ok (anything mutable would work), but nevermind...